src/gui/osutils/nixutils/DeviceListenerLibUsb.cpp

```diff
@@ -84,7 +84,7 @@ int DeviceListenerLibUsb::registerHotplugCallback(bool arrived, bool left, int v
             return 0;
         },
         that,
-        &handle);
+        &static_cast<libusb_hotplug_callback_handle>(handle));
```

Comment:
I think your runtime error is here, you shouldn't reference the output of the static cast directly. Make the handle variable an int, then static cast the return call.

Reply:
Why int?

Please keep in mind that I'm trying to fix this in FreeBSD. FreeBSD has its own implementation of libusb. AFAIK other BSDs do too, but I have not checked.

FreeBSD's `libusb_hotplug_callback_handle` is defined as [1]:

```c
typedef struct libusb_hotplug_callback_handle_struct *libusb_hotplug_callback_handle;
```

so the error I'm getting is:

```
/wrkdirs/usr/ports/security/keepassxc/work/keepassxc-2.7.8/src/gui/osutils/nixutils/DeviceListenerLibUsb.cpp:87:10: error: cannot cast from type 'Handle' (aka 'long long') to pointer type 'libusb_hotplug_callback_handle' (aka 'libusb_hotplug_callback_handle_struct *')
   87 |         &static_cast<libusb_hotplug_callback_handle>(handle));
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/wrkdirs/usr/ports/security/keepassxc/work/keepassxc-2.7.8/src/gui/osutils/nixutils/DeviceListenerLibUsb.cpp:110:77: error: cannot cast from type 'Handle' (aka 'long long') to pointer type 'libusb_hotplug_callback_handle' (aka 'libusb_hotplug_callback_handle_struct *')
  110 |     libusb_hotplug_deregister_callback(static_cast<libusb_context*>(m_ctx), static_cast<libusb_hotplug_callback_handle>(handle));
      |                                                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 errors generated.
```

I'm not sure what I should cast to at this point, on FreeBSD.

If later Linux (or any other OS) requires a different code I will use `#ifdef`s to differentiate.


[1] https://github.com/freebsd/freebsd-src/blob/48edad2edf6eb7a539e40dad8e1f87e3fa4973fd/lib/libusb/libusb.h#L311

Reply:
Either way your static cast is incorrect in this context. Slow clap for FreeBSD for a completely divergent API.

Reply:
Well I don't know how and why a different API evolved in FreeBSD, but I'm trying to understand why a 64bit number on a 64bit arch cannot be cast to a pointer. The memory representation of them is the same AFAIK.

What I'm trying to do is adapt the code in keepassxc for these function to pass around pointers in place of integers. Apart fro the different type (with a different size in most cases), the API is not that much different, so the basic concept should be possible to put in practice.

Reply:
Could be a libusb version thing. On Ubuntu (libusb 1.0.26), `libusb_hotplug_callback_handle` is typedef'ed to `int`. FreeBSD has an entirely different implementation of the whole library.

I think the main problem is that `libusb_hotplug_callback_handle` itself is not a pointer type on FreeBSD. Instead, they chose to typedef only `*libusb_hotplug_callback_handle`, so you have to use it with `*`.

Reply:
@phoerious

Looks like FreeBSD has an old implementation of libusb included in the base system. This will be addressed at some point, but will require time.

I can try a different patch, but since I do not own a ubikey I can properly test them myself, and getting feedback requires time.